pkg/resource/node-resource-manager.go

```diff
@@ -0,0 +1,233 @@
+
+	stateChann chan map[string][]common.TimeSeries
+
+	//TODO: use state to be a backup of tsp
+	// A copy of data from stateChann
+	state map[string][]common.TimeSeries
+	// Updated when get new data from stateChann, used to determine whether state has expired
+	lastStateTime time.Time
+}
+
+func NewNodeResourceManager(client clientset.Interface, nodeName string, podInformer coreinformers.PodInformer, nodeInformer coreinformers.NodeInformer,
+	tspInformer predictionv1.TimeSeriesPredictionInformer, runtimeEndpoint string, stateChann chan map[string][]common.TimeSeries, Collectors *sync.Map) *NodeResourceManager {
+	o := &NodeResourceManager{
+		nodeName:   nodeName,
+		client:     client,
+		nodeLister: nodeInformer.Lister(),
+		nodeSynced: nodeInformer.Informer().HasSynced,
+		tspLister:  tspInformer.Lister(),
+		tspSynced:  tspInformer.Informer().HasSynced,
+		stateChann: stateChann,
+	}
+	return o
+}
+
+func (o *NodeResourceManager) Name() string {
+	return "NodeResourceManager"
+}
+
+func (o *NodeResourceManager) Run(stop <-chan struct{}) {
+	klog.Infof("Starting node resource manager.")
+
+	// Wait for the caches to be synced before starting workers
+	if !cache.WaitForNamedCacheSync("node-resource-manager",
+		stop,
+		o.tspSynced,
+		o.nodeSynced,
+	) {
+		return
+	}
+
+	go func() {
+		tspUpdateTicker := time.NewTicker(TspUpdateInterval)
+		defer tspUpdateTicker.Stop()
+		for {
+			select {
+			case state := <-o.stateChann:
+				o.state = state
+				o.lastStateTime = time.Now()
+			case <-tspUpdateTicker.C:
+				o.UpdateNodeResource()
+			case <-stop:
+				klog.Infof("node resource manager exit")
+				return
+			}
+		}
+	}()
+
+	return
+}
+
+func (o *NodeResourceManager) UpdateNodeResource() {
+	tsps, err := o.tspLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("List tsp error: %#v", err)
+		return
+	}
+
+	node, addresses := o.getNode()
+	if len(addresses) == 0 {
+		klog.Error("Node addresses is empty")
+		return
+	}
+	nodeCopy := node.DeepCopy()
+
+	for _, tsp := range tsps {
+		// Get current node info
+		target := tsp.Spec.TargetRef
+		if target.Kind != config.TargetKindNode {
+			return
+		}
+
+		// Whether tsp is matched with this node
+		tspMatched, err := o.FindTargetNode(tsp, addresses)
+		if err != nil {
+			klog.Error(err.Error())
+		}
+
+		if !tspMatched {
+			return
+		}
+
+		o.BuildNodeStatus(tsp, nodeCopy)
+		if !equality.Semantic.DeepEqual(&node.Status, &nodeCopy.Status) {
+			// Update Node status extend-resource info
+			// TODO fix: strategic merge patch kubernetes
+			if _, err := o.client.CoreV1().Nodes().Update(context.TODO(), nodeCopy, metav1.UpdateOptions{}); err != nil {
+				klog.Errorf("Failed to update node %s's status extend-resource, %v", nodeCopy.Name, err)
+				return
+			}
+			klog.V(4).Infof("Update Node %s Extend Resource Success according to TSP %s", node.Name, tsp.Name)
+		}
+		return
+	}
+}
+
+func (o *NodeResourceManager) getNode() (*v1.Node, []v1.NodeAddress) {
```

Comment:
Not return NodeAddress here, we can get address from the node struct

Reply:
Ok